File: #0712.minimum-ascii-delete-sum-for-two-strings.cpp
class Solution {
public:
    int minimumDeleteSum(string s1, string s2) {

        int size1 = s1.size();
        int size2 = s2.size();
        vector<vector<int>> result(size1 + 1, vector(size2 + 1, 0));
        for (int i = 0; i <= size1; i++) {
            for (int j = 0; j <= size2; j++) {
                if (i == 0 && j == 0){
                    result[i][j] = 0;
                } else if(j == 0) {
                    result[i][j] = result[i - 1][j] + s1[i - 1];
                } else if (i == 0) {
                    result[i][j] = result[i][j - 1] + s2[j - 1];
                } else if (s1[i-1] == s2[j - 1]) {
                    result[i][j] = result[i - 1][j - 1];
                } else {
                    result[i][j] = std::min(
                        result[i - 1][j] + s1[i - 1],
                        result[i][j - 1] + s2[j - 1]
                    );
                }
            }
        }
        return result[size1][size2];
    }
};

File: #0712.minimum-ascii-delete-sum-for-two-strings.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
var minimumDeleteSum = function(s1, s2) {
    let dp = [
        ...Array(s1.length + 1),
    ].map((_) => Array(s2.length + 1));
    dp[0][0] = 0;
	
    for (let j = 1; j <= s2.length; j++) {
        dp[0][j] = dp[0][j - 1] + s2.charCodeAt(j - 1);
    }

    for (let i = 1; i <= s1.length; i++) {
        dp[i][0] = dp[i - 1][0] + s1.charCodeAt(i - 1);
    }

    for (let i = 1; i <= s1.length; i++) {
        for (let j = 1; j <= s2.length; j++) {
            if (s1.charAt(i - 1) === s2.charAt(j - 1)) {
                dp[i][j] = dp[i - 1][j - 1];
            } else {
                 dp[i][j] = Math.min(dp[i - 1][j] + s1.charCodeAt(i - 1),
                    dp[i][j - 1] + s2.charCodeAt(j - 1));
            }
        }
    }
    return dp[s1.length][s2.length];
};
